webhooks/clusterclass.go

```diff
@@ -123,8 +140,126 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
+	mdcRecords := map[string]string{}
+	for _, c := range clusters {
+		for _, machineDeploymentClass := range c.Spec.Topology.Workers.MachineDeployments {
+			if missingClasses.Has(machineDeploymentClass.Class) {
+				mdcRecords[fmt.Sprintf("%v/%v", c.Name, machineDeploymentClass.Name)] = machineDeploymentClass.Class
+			}
+		}
+	}
+
+	// For each machineDeployment using a managed topology return an error if it is using a class that has been removed in the ClusterClass change.
+	for _, md := range machineDeployments {
+		mdName := fmt.Sprintf("%v/%v", md.Labels[clusterv1.ClusterLabelName], md.Labels[clusterv1.ClusterTopologyMachineDeploymentLabelName])
+		if _, ok := mdcRecords[mdName]; ok {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath(""), fmt.Sprintf("MachineDeploymentClass %v is in use in MachineDeployment %v in Cluster %v. ClusterClass %v modification not allowed",
+				mdcRecords[mdName], md.Labels[clusterv1.ClusterTopologyMachineDeploymentLabelName], md.Labels[clusterv1.ClusterLabelName], old.Name),
+			))
+		}
+	}
+
+	return allErrs
+}
+
+func (webhook *ClusterClass) topologyManagedMachineDeployments(namespace string) ([]clusterv1.MachineDeployment, field.ErrorList) {
+	var allErrs field.ErrorList
+
+	// List all the MachineDeployments in the current cluster using a managed topology.
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	err := webhook.Client.List(context.Background(), machineDeployments,
+		client.MatchingLabels{
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		// TODO: Should this be namespaced in future?
+		client.InNamespace(namespace),
+	)
+	if err != nil {
+		allErrs = append(allErrs, field.InternalError(field.NewPath(""), err))
+		return nil, allErrs
+	}
+	return machineDeployments.Items, nil
+}
+
+func (webhook *ClusterClass) removedMachineClasses(old, new *clusterv1.ClusterClass) sets.String {
+	missingClasses := sets.NewString()
+
+	// Ensure no MachineDeploymentClass in use has been removed.
+	classes := webhook.classNamesFromWorkerClass(new.Spec.Workers)
+	for _, oldClass := range old.Spec.Workers.MachineDeployments {
+		if !classes.Has(oldClass.Class) {
+			missingClasses.Insert(oldClass.Class)
+		}
+	}
+	return missingClasses
+}
+
+// classNames returns the set of MachineDeployment class names.
+func (webhook *ClusterClass) classNamesFromWorkerClass(w clusterv1.WorkersClass) sets.String {
+	classes := sets.NewString()
+	for _, class := range w.MachineDeployments {
+		classes.Insert(class.Class)
+	}
+	return classes
+}
+
+func (webhook *ClusterClass) clustersUsingClusterClass(clusterClass *clusterv1.ClusterClass) ([]clusterv1.Cluster, field.ErrorList) {
+	var allErrs field.ErrorList
+	clusters := &clusterv1.ClusterList{}
+	clustersUsingClusterClass := []clusterv1.Cluster{}
+	err := webhook.Client.List(context.Background(), clusters,
+		client.MatchingLabels{
+			clusterv1.ClusterTopologyOwnedLabel: "",
+		},
+		//TODO: This will be an issue when doing clusterClass across namespaces.
```

Comment:
Not sure if we should have that TODO given that we have to look at a lot of places for multi-ns support and I'm not sure it's sustainable/ makes sense to already add TODOs at all relevant places.

Reply:
Yeah - there's an issue to modify the way namespacing for clusterclass works #5673. I just wanted to call out the parts of this code that will be impacted.

Reply:
Got it. I would remove them on this PR, but up to you.

Reply:
But it might be a nice idea to add a comment to the related issue about that dependency :)